synapseclient/table.py

```diff
@@ -1811,17 +1811,25 @@ def setColumnHeaders(self, headers):
 
     def __iter__(self):
         def iterate_rows(filepath, headers):
+            if not self.header or not self.headers:
+                raise ValueError("Iteration not supported for table without headers.")
+
+            header_name = {header.name for header in headers}
+            row_metadata_headers = {'ROW_ID', 'ROW_VERSION', 'ROW_ETAG'}
+            num_row_metadata_in_headers = len(header_name & row_metadata_headers)
             with io.open(filepath, encoding='utf-8', newline=self.lineEnd) as f:
                 reader = csv.reader(f,
                     delimiter=self.separator,
                     escapechar=self.escapeCharacter,
                     lineterminator=self.lineEnd,
                     quotechar=self.quoteCharacter)
-                if self.header:
-                    header = next(reader)
-
-                for row in reader:
-                    yield cast_values(row, headers)
+                csv_header = set(next(reader))
+                num_metadata_cols_diff = len(csv_header & row_metadata_headers) - num_row_metadata_in_headers
+                if num_metadata_cols_diff == 0 or (num_metadata_cols_diff > 0 and num_row_metadata_in_headers == 0):
```

Comment:
One thing I noticed during the reading the logic of the code was that I was converting booleans related to `num_metadata_cols_diff ` into the relationship between `num_row_metadata_in_headers` and `len(csv_header & row_metadata_headers)` (i.e. `num_row_metadata_columns_in_csv`).

In other words, the `num_metadata_cols_diff ` doesn't immediately relay what being `>0` or `<0` means; I must reread the variable declaration of the `num_metadata_cols_diff ` to figure it out. I feel it is easier to read/understand when you use `num_row_metadata_columns_in_csv > num_row_metadata_in_headers`. I would reserve the use of the `num_metadata_cols_diff ` only for `cast_values(row[num_metadata_cols_diff:], headers)`